examples/dft.masm

```diff
@@ -0,0 +1,288 @@
+        dup.1                                 # stack: [i+1, n, i+1, n, v{i+1}, ..., v{n-1}]
+        neq                                   # stack: [1 or 0, i+1, n, v{i+1}, ..., v{n-1}]
+    end
+
+    # stack: [n, n, ...]
+    drop                                      # stack: [n, ...]
+    drop                                      # stack: [...]
+end
+
+#! Retrieves an array previously stored with `proc.store_array`.
+#! Input: [array_loc], the memory location to retrieve the array
+#! Output: [n, v0, ..., v{n-1}]
+#!
+#! Locals
+#! 0: array_loc
+proc.retrieve_array.1
+    dup                                       # stack: [array_loc, array_loc, ...]
+    loc_store.0                               # stack: [array_loc, ...]
+    mem_load                                  # stack: [n, ...]
+
+    # Prepare stack for loop
+    push.0.1                                  # pushes i=0, followed by 1 so that we enter the loop
+
+    while.true
+        # stack: [i, n, v{i}, ..., v{n-1}]
+
+        # Setup read offset
+        # Note: we read at location `array_loc + n - i`, since the element at array_loc is `n`
+        # (i.e. the array starts at address `array_loc + 1`)
+        dup.1                                   # stack: [n, i, n, v{i+1}, ..., v{n-1}]
+        dup.1                                   # stack: [i, n, i, n, v{i+1}, ..., v{n-1}]
+        sub                                     # stack: [n - i, i, n, v{i+1}, ..., v{n-1}]
+        loc_load.0                              # stack: [array_loc, n - i, i, n, v{i+1}, ..., v{n-1}]
+        add                                     # stack: [array_loc + n - i, i, n, v{i+1}, ..., v{n-1}]
+
+        # Read v{i} and put in its spot on the stack
+        mem_load                                # stack: [v{i}, i, n, v{i+1}, ..., v{n-1}]
+        movdn.2                                 # stack: [i, n, v{i}, v{i+1}, ..., v{n-1}]
+
+        # Increment i and check if we're done
+        add.1                                 # stack: [i+1, n, v{i}, v{i+1}, ..., v{n-1}]
+        dup.1                                 # stack: [n, i+1, n, v{i}, v{i+1}, ..., v{n-1}]
+        dup.1                                 # stack: [i+1, n, i+1, n, v{i}, v{i+1}, ..., v{n-1}]
+        neq                                   # stack: [1 or 0, i+1, n, v{i}, v{i+1}, ..., v{n-1}]
+    end
+
+    # stack: [n, n, v0, ..., v{n-1}]
+
+    drop                                      # stack: [n, v0, ..., v{n-1}]
+end
+
+#! Input: [n, v0, ..., v{n-1}]
+#!
+#! Locals
+#! 0: n
+#! 1: root_of_unity
+#! 2: k
+proc.main.3
+    # Initialize memory
+    # Note: k is initialized to 0 by the VM
+    dup                                      # stack: [n, n, v0, ..., v{n-1}]
+    loc_store.0                              # stack: [n, v0, ..., v{n-1}]
+    dup                                      # stack: [n, n, v0, ..., v{n-1}]
+    mem_store.RESULT_ARR_LOC                 # stack: [n, v0, ..., v{n-1}]
```

Comment:
I was trying to keep the semantics separate:
- `local[0]` stores the length of the array
- `RESULT_ARR_LOC` stores the result array, which happens to be encoded as `[length of array, ele_0, ele_1, ...]`

This allows us to change the encoding the of the final array without creating mysterious bugs, as well as making it more readable (i.e. we read `loc_load.0` to load the length of the array instead of `loc_load.RESULT_ARR_LOC`)